src/commonMain/kotlin/fr/acinq/lightning/wire/LiquidityAds.kt

```diff
@@ -20,140 +24,352 @@ import fr.acinq.lightning.utils.sat
+                        )
+                    )
+                }
+            }
+        }
+
+        fun write(out: Output) = when (this) {
+            is Basic -> {
+                val encoded = encode()
+                LightningCodecs.writeBigSize(0, out) // tag
+                LightningCodecs.writeBigSize(encoded.size.toLong(), out) // length
+                LightningCodecs.writeBytes(encoded, out)
+            }
+        }
+
+        companion object {
+            fun read(input: Input): FundingLease? {
+                val tag = LightningCodecs.bigSize(input)
+                // We always read the lease data, even if we don't support it: this lets us skip unknown lease types.
+                val content = LightningCodecs.bytes(input, LightningCodecs.bigSize(input))
+                return when (tag) {
+                    0L -> Basic.decode(content)
+                    else -> null
+                }
+            }
+        }
+    }
+
+    sealed class FundingLeaseWitness {
+        abstract val fundingScript: ByteVector
+
+        /** The seller signs a witness that commits to the funding script. */
+        data class Basic(override val fundingScript: ByteVector) : FundingLeaseWitness()
+
+        fun signedData(): ByteArray {
+            val tag = when (this) {
+                is Basic -> "basic_funding_lease"
+            }
+            val tmp = ByteArrayOutput()
+            write(tmp)
+            return Crypto.sha256(tag.encodeToByteArray() + tmp.toByteArray())
+        }
+
+        fun write(out: Output) = when (this) {
+            is Basic -> {
+                LightningCodecs.writeBigSize(0, out) // tag
+                LightningCodecs.writeBigSize(fundingScript.size().toLong(), out) // length
+                LightningCodecs.writeBytes(fundingScript, out)
+            }
+        }
+
+        companion object {
+            fun read(input: Input): FundingLeaseWitness = when (val tag = LightningCodecs.bigSize(input)) {
+                0L -> Basic(LightningCodecs.bytes(input, LightningCodecs.bigSize(input)).byteVector())
+                else -> throw IllegalArgumentException("unknown funding lease witness (tag=$tag)")
+            }
+        }
+    }
+
+    /** The fees associated with a given [LeaseRate] can be paid using various options. */
+    sealed class PaymentType {
+        /** Fees are transferred from the buyer's channel balance to the seller's during the interactive-tx construction. */
+        data object FromChannelBalance : PaymentType()
+        /** Fees will be deducted from future HTLCs that will be relayed to the buyer. */
+        data object FromFutureHtlc : PaymentType()
+        /** Fees will be deducted from future HTLCs that will be relayed to the buyer, but the preimage is revealed immediately. */
+        data object FromFutureHtlcWithPreimage : PaymentType()
+        /** Fees will be taken from the buyer's current fee credit. */
+        data object FromFeeCredit : PaymentType()
```

Comment:
It should not be an either/or between `FromFeeCredit` on one hand, and `FromChannelBalance `/`FromFutureHtlc` on the other hand. Quite often the fee credit will only cover part of the channel management fee.

Reply:
> We would *not* create a `from_fee_credit` payment type and would instead implicitly use it

That is much closer to the current implementation

> unless the LSP always consumes the whole channel balance and then takes the rest from fee credit?

We need to spend first from the fee credit, so it always converges to zero.

> or the LSP sends some data in splice_ack to let the wallet know how much will be taken from fee credit and how much from channel balance?

The LSP could provide the current fee credit as a tlv in `splice_ack`. Is there a scenario where the fee credit would be updated while a splice is in progress?

> the LSP uses the fee credit in priority, and deduces the remaining amount from the future HTLCs

That is safer for the LSP than the non-fee-credit variant because the buyer has something at stake.

Reply:
SGTM, feel free to juste ignore the last commit for now, I'll extract it to a separate PR during my next rebase and will rework it to match that. But I'll focus on finalizing the eclair implementation of the on-the-fly funding part first, and will come back to fee credit afterwards.

Reply:
I moved fee credit to #660 and will re-work it once I'm done with the `eclair` side.